lib/src/rules/avoid_annotating_parameters_with_function.dart

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.avoid_annotating_parameters_with_function;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const _desc = r'Avoid annotating parameters with Function.';
+
+const _details = r'''
+
+**AVOID** annotating parameters with Function.
+
+**BAD:**
+```
+bool isValidString(String value, Function predicate) {
+  ...
+}
+```
+
+**GOOD:**
+```
+bool isValidString(String value, bool predicate(String string)) {
+  ...
+}
+```
+
+''';
+
+class AvoidAnnotatingParametersWithFunction extends LintRule {
+  _Visitor _visitor;
+  AvoidAnnotatingParametersWithFunction()
+      : super(
+            name: 'avoid_annotating_parameters_with_function',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitSimpleFormalParameter(SimpleFormalParameter node) {
+    if (DartTypeUtilities.isClass(node.type.type, 'Function', 'dart.core')) {
```

Comment:
This doesn't account for the exception given in the guideline. It's worth thinking about whether we can detect that case.